lib/srv/server/installer/autodiscover.go

```diff
@@ -0,0 +1,600 @@
+	}
+
+	teleportNodeConfigureArgs := []string{"node", "configure", "--output=file://" + teleportYamlConfigurationPath,
+		fmt.Sprintf(`--proxy=%s`, shsprintf.EscapeDefaultContext(asi.ProxyPublicAddr)),
+		fmt.Sprintf(`--join-method=%s`, shsprintf.EscapeDefaultContext(string(joinMethod))),
+		fmt.Sprintf(`--token=%s`, shsprintf.EscapeDefaultContext(asi.TokenName)),
+		fmt.Sprintf(`--labels=%s`, shsprintf.EscapeDefaultContext(nodeLabelsCommaSeperated)),
+	}
+	if asi.AzureClientID != "" {
+		teleportNodeConfigureArgs = append(teleportNodeConfigureArgs,
+			fmt.Sprintf(`--azure-client-id=%s`, shsprintf.EscapeDefaultContext(asi.AzureClientID)))
+	}
+
+	asi.Logger.InfoContext(ctx, "Writing teleport configuration", "teleport", asi.binariesLocation.teleport, "args", teleportNodeConfigureArgs)
+	teleportNodeConfigureCmd := exec.CommandContext(ctx, asi.binariesLocation.teleport, teleportNodeConfigureArgs...)
+	teleportNodeConfigureCmdOutput, err := teleportNodeConfigureCmd.CombinedOutput()
+	if err != nil {
+		return trace.Wrap(err, string(teleportNodeConfigureCmdOutput))
+	}
+
+	asi.Logger.InfoContext(ctx, "Enabling and starting teleport service")
+	systemctlEnableNowCMD := exec.CommandContext(ctx, asi.binariesLocation.systemctl, "enable", "--now", "teleport")
+	systemctlEnableNowCMDOutput, err := systemctlEnableNowCMD.CombinedOutput()
+	if err != nil {
+		return trace.Wrap(err, string(systemctlEnableNowCMDOutput))
+	}
+
+	return nil
+}
+
+func (asi *AutodiscoverNodeInstaller) getIMDSClient(ctx context.Context) (imds.Client, error) {
+	// detect and fetch cloud provider metadata
+	imdsClient, err := cloud.DiscoverInstanceMetadata(ctx, asi.imdsProviders)
+	if err != nil {
+		if trace.IsNotFound(err) {
+			return nil, trace.BadParameter("Auto Discover only runs on Cloud instances with IMDS/Metadata service enabled. Ensure the service is running and try again.")
+		}
+		return nil, trace.Wrap(err)
+	}
+
+	return imdsClient, nil
+}
+
+func (asi *AutodiscoverNodeInstaller) fetchTargetVersion(ctx context.Context) string {
+	upgradeURL, err := url.Parse(asi.autoUpgradesChannelURL)
+	if err != nil {
+		asi.Logger.WarnContext(ctx, "Failed to parse automatic upgrades default channel url, using api version",
+			"channel_url", asi.autoUpgradesChannelURL,
+			"error", err,
+			"version", api.Version)
+		return api.Version
+	}
+
+	autoUpgradesVersionGetter := version.NewBasicHTTPVersionGetter(upgradeURL)
+	targetVersion, err := autoUpgradesVersionGetter.GetVersion(ctx)
+	if err != nil {
+		asi.Logger.WarnContext(ctx, "Failed to query target version, using api version",
+			"error", err,
+			"version", api.Version)
+		return api.Version
+	}
+	asi.Logger.InfoContext(ctx, "Found target version",
+		"channel_url", asi.autoUpgradesChannelURL,
+		"version", targetVersion)
+
+	return strings.TrimSpace(strings.TrimPrefix(targetVersion, "v"))
+}
+
+func fetchNodeAutoDiscoverLabels(ctx context.Context, imdsClient imds.Client) (map[string]string, error) {
```

Comment:
I think we already automatically fetch node labels on every cloud we support. Is this something we can reuse? cc @atburke would probably know.

Reply:
For the record, this is the equivalent in the current shell script
https://github.com/gravitational/teleport/blob/d9bcd2e2dafd64b05331734306b882c2855de97c/api/types/installers/installer.sh.tmpl#L151-L183

These labels are not sourced from the instance's labels/tags. These are mostly metadata used for detecting already enrolled instances and some other metadata.

Let's use Azure as an example.
The script adds the following labels:
https://github.com/gravitational/teleport/blob/d9bcd2e2dafd64b05331734306b882c2855de97c/api/types/installers/installer.sh.tmpl#L155-L161

The following labels are used to detect if the instance is already enrolled:
https://github.com/gravitational/teleport/blob/8410d1e4fb6d5e1400f79de337a8da5c2b1de4cd/lib/srv/discovery/discovery.go#L1046-L1049

So, we end up with a couple more labels which might be used for RBAC.
Some of those labels are not present in any go code, and that's why I ended up adding them here
https://github.com/gravitational/teleport/blob/7e35c38679a71601234305baffa3cfa83c5b5fa3/api/types/constants.go#L675-L682

@atburke Can you please double check this?

Reply:
Yep, that's all correct.